lib/mayaUsd/fileio/translators/translatorMayaReference.cpp

```diff
@@ -424,29 +472,74 @@ MStatus UsdMayaTranslatorMayaReference::update(const UsdPrim& prim, MObject pare
+                        if (!tempRefFn.isFromReferencedFile()) {
+                            if (expectedValue == tempRefFn.name()) {
+                                // Reconnect the reference node's `associatedNode` attr before
+                                // loading it, since the previous connection may be gone.
+                                connectReferenceAssociatedNode(parentDag, tempRefFn);
+                                refNode = tempRefNode;
+
+                                // Update Maya Ref Node on reconnect
+                                MString uniqueRefNodeName
+                                    = getUniqueRefNodeName(prim, parentDag, tempRefFn);
+                                tempRefFn.setName(uniqueRefNodeName);
+
+                                MString refDependNodeName = tempRefFn.name();
+                                VtValue value(std::string(
+                                    refDependNodeName.asChar(), refDependNodeName.length()));
+                                prim.SetCustomDataByKey(maya_associatedReferenceNode, value);
+
+                                if (prim.HasAttribute(MayaReferenceAttribute)) {
+                                    UsdAttribute attr = prim.GetAttribute(MayaReferenceAttribute);
+                                    attr.Set(refDependNodeName.asChar());
+                                }
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+        } else {
+            // Old behaviour for ALMayaReference
+            for (MItDependencyNodes refIter(MFn::kReference); !refIter.isDone(); refIter.next()) {
+                MObject      tempRefNode = refIter.item();
+                MFnReference tempRefFn(tempRefNode);
+                if (!tempRefFn.isFromReferencedFile()) {
+
+                    // Get the name of the reference node so we can check if this node
+                    // matches the prim we are processing.  Strip off the "_RN" suffix.
+                    //
+                    MString refNodeName = tempRefFn.name();
+                    MString refName(refNodeName.asChar(), refNodeName.numChars() - 3);
+
+                    // What reference node name is the prim expecting?
+                    MString expectedRefName = refNameFromPath(parentDag);
+
+                    // If found a match, reconnect the reference node's `associatedNode`
+                    // attr before loading it, since the previous connection may be gone.
+                    //
+                    if (refName == expectedRefName) {
+                        // Reconnect the reference node's `associatedNode` attr before
+                        // loading it, since the previous connection may be gone.
+                        connectReferenceAssociatedNode(parentDag, tempRefFn);
+                        refNode = tempRefNode;
+                        break;
+                    }
```

Comment:
I would sort of consider this old behavior to be "broken". I fixed it in this PR https://github.com/Autodesk/maya-usd/pull/1620 using the reference node name strategy, but I believe the "session" layer storage strategy should be even less-obtrusive. I think we can fix old and new for the same issues!